added ability to auto detect .vercel/project.json files

aarondill opened the pull request:
# What it does:
Allows this extension to automatically detect project ids from .vercel/project.json files if there is no setting defined. This allows for a global override to be set in (global)settings.json, or a workspace override to be set by creating a .vscode/settings.json file, just like before, but in the absence of either value, checks for a top level .vercel/project.json file (created by the Vercel CLI) to get the project id. 
### Reason for creation
Allows for more dynamic usage with the Vercel CLI(likely to already be in use by anyone utilizing this extension), and avoids unnecessary user effort in defining a project name in a seperate file for every Vercel app they edit.

### Shortfall
This would not be supported with projects that have their .vercel folder nested within another, however such users could set a value in their workspace settings file.

### What it solves: 
See issue #17

haydenbleasel commented on package.json:
```diff
@@ -70,7 +70,8 @@
     "stylelint": "^14.14.1",
-    "stylelint-prettier": "^2.0.0"
+    "stylelint-prettier": "^2.0.0",
+    "yarn": "^1.22.19"
```
Probably don't need yarn as a dependency.

haydenbleasel commented on package.json:
```diff
@@ -70,7 +70,8 @@
     "stylelint": "^14.14.1",
-    "stylelint-prettier": "^2.0.0"
+    "stylelint-prettier": "^2.0.0",
+    "yarn": "^1.22.19"
```
Probably don't need yarn as a dependency.

aarondill replied:
Yep, sorry, Installed that to run the tests. Will remove ASAP

haydenbleasel commented on src/extension.ts:
```diff
@@ -11,29 +11,56 @@ const toSentenceCase = (str: string): string => {
 };
 
+const getProjectIdFromJson = async (): Promise<string | undefined> => {
+  if (workspace.workspaceFolders?.[0] === undefined) {
```
Would `(!workspace.workspaceFolders?.[0])` suffice?

haydenbleasel commented on src/extension.ts:
```diff
@@ -11,29 +11,56 @@ const toSentenceCase = (str: string): string => {
+  try {
+    const stringJson: string = Buffer.from(vercelProjectJson).toString('utf8');
+    const parsedVercelJSON: { projectId: string } = JSON.parse(stringJson) as {
+      projectId: string;
```
We should make this `projectId?: string` and check for `projectId` as there's no assurance that it definitely exists.

haydenbleasel replied:
We return `String(parsedVercelJSON.projectId)` on the next line. If the `projectId` is undefined, then `String(undefined) -> 'undefined'` which will pass a string check.

aarondill replied:
I believe I fixed this issue. I am still new to using typescript, so a second set of eyes would be greatly appreciated.

haydenbleasel reviewed (changes requested):
Hey @aarondill, thanks for the awesome enhancement! Can you check out these 3 items? Once they're resolved, I'm keen to merge.

haydenbleasel commented:
One more thing! We'll need an update to the README saying we have `.vercel/project.json` support now :)

aarondill commented:
@haydenbleasel, Should the extension give a prompt if the file is not found? Additionally, should it give a prompt if the setting is not found? One or the other would be preferable to reduce the number of user notifications. Additionally, there could be a prompt if neither option is found, simply stating as such; or there could be no prompt, allowing the extension to work silently in the background if no vercel app is configured for the workspace.
Being your extension, that is your design choice to make, but my personal preference would be the silent execution which only bothers the user if an *error* occurs.

aarondill commented:
Created updated README.md File.
I just saw your recent commit that creates utilities. I will rewrite this for that. Out of curiosity, would you prefer that this function were a seperate utility, or that I continue with the inline nature it currently is?

aarondill commented:
> Created updated README.md File. I just saw your recent commit that creates utilities. I will rewrite this for that. Out of curiosity, would you prefer that this function were a seperate utility, or that I continue with the inline nature it currently is?

Nevermind, this function requires too many of the imports to make it reasonable(IMO) to seperate it. You are welcome to let me know if you disagree.

aarondill commented:
I have updated the code to fix the mistakes pointed out (though like i said, I would love a second set of eyes on that type checking).
Additionally, I have updated the README.md
The code now merges with no conflicts.
## Still remaining:
I would like to know your decision on the notifying the user issue(see above). Once you let me know your judgment, I will modify it(if necessary) and then this *should* be ready to merge.

haydenbleasel commented:
> @haydenbleasel, Should the extension give a prompt if the file is not found? Additionally, should it give a prompt if the setting is not found? One or the other would be preferable to reduce the number of user notifications. Additionally, there could be a prompt if neither option is found, simply stating as such; or there could be no prompt, allowing the extension to work silently in the background if no vercel app is configured for the workspace. Being your extension, that is your design choice to make, but my personal preference would be the silent execution which only bothers the user if an _error_ occurs.

That's a great callout. Preferably it should fail silently as you mentioned. I'm happy to solve this in a different MR.

aarondill commented:
This change should be as simple as removing this line on line 18:
```diff
    let vercelProjectJson: Uint8Array | null = null;
   try {
     vercelProjectJson = await workspace.fs.readFile(fileUri);
   } catch {
-     await window.showErrorMessage('Could not find .vercel/project.json file.');
     return undefined;
   }
```
My commit has already remove the error for not setting the configuration, and you should still keep the one requesting an access key, though, perhaps the code might be rethought slightly to only check for an access token after determining that a project id is present, so that users could set individual access tokens in workspaces and not get errors in other workspaces where vercel is not utilized. 
I don’t currently have time to make these changes, however in the next couple days I could do that

aarondill commented:
That could be done simply by moving lines 37 and 41 through 46 to line 59(after determining that a project id is available)

aarondill commented:
Apparently you can change files directly on github!
Anyways, I have made the appropriate changes and recommited.
**WARNING:** I have *not* tested the last commit. It should still work, but perhaps clone it and compile it before merging...
